src/wallet/api.rs

```diff
@@ -1060,48 +1074,172 @@ impl Wallet {
-            .filter(|(utxo, spend_info)| {
+            .filter(|(utxo, _)| {
                 let outpoint = OutPoint::new(utxo.txid, utxo.vout);
                 !locked_utxos.contains(&outpoint)
-                    && !matches!(spend_info, UTXOSpendInfo::FidelityBondCoin { .. })
             })
             .collect::<Vec<_>>();
 
-        // Sort in descending order so that we use the largest UTXOs first.
-        unspents.sort_by(|a, b| b.0.amount.cmp(&a.0.amount));
+        // Assertion statement : fidelity coins are not included
+        assert!(
+            unspents.iter().all(|(_, spend_info)| !matches!(
+                spend_info,
+                UTXOSpendInfo::FidelityBondCoin { .. }
+            )),
+            "Fidelity coins are not included in coin selection"
+        );
 
-        let mut selected_utxo = Vec::new();
-        let mut remaining = amount;
+        // Get next external and internal addresses using dedicated wallet APIs
+        let external_address = self.get_next_external_address()?;
+        let internal_address = self.get_next_internal_addresses(1)?[0].clone();
 
-        // Simple largest-first coin selection.
-        for unspent in unspents {
-            if remaining.checked_sub(unspent.0.amount).is_none() {
-                selected_utxo.push(unspent);
-                break;
-            } else {
-                remaining -= unspent.0.amount;
-                selected_utxo.push(unspent);
+        // TODO : Add support for addresses other than P2PKH
+        let external_address_script_pubkey = ScriptBuf::new_p2pkh(
+            &external_address
+                .pubkey_hash()
+                .ok_or_else(|| WalletError::General("Could not get pubkey hash".to_string()))?,
+        );
+
+        let internal_address_script_pubkey =
+            ScriptBuf::new_p2pkh(&internal_address.pubkey_hash().ok_or_else(|| {
+                WalletError::General("Could not get change pubkey hash".to_string())
+            })?);
+
+        // Prepare CoinSelectionOpt: Calculate required weights and costs, additional metrics
+        const LONG_TERM_FEERATE: f32 = 10.0;
+        let change_weight = Weight::from_vb_unwrap(
+            (Amount::SIZE // 8 bytes for amount
+                    + VarInt::from(internal_address_script_pubkey.len()).size() // VarInt size for script_pubkey
+                    + internal_address_script_pubkey.len()) as u64, // script_pubkey size
+        );
+        // Total cost associated with creating and later spending a change output in a transaction.
+        // This includes the transaction fees for both the current transaction (where the change is created) and the future transaction (where the change is spent).
+        // Calculate fee generally uses the units :- vbytes * sats/vbyte
+        let cost_of_change = calculate_fee(change_weight.to_vbytes_ceil(), LONG_TERM_FEERATE)
+            .map_err(|e| WalletError::General(format!("Fee calculation failed: {}", e)))?;
```

Comment:
What you need is `change_weight * fee_rate + spending_input_weight * long_term_feerate`